lib/queuePopulator/IngestionReader.js

```diff
@@ -0,0 +1,233 @@
+                const offset = Number.parseInt(res, 10);
+                if (err) {
+                    return next(err);
+                }
+                if (offset === 1) {
+                    this.newIngestion = true;
+                }
+                return next();
+            }),
+            next => {
+                if (!this.newIngestion) {
+                    return next();
+                }
+                return this._iProducer.snapshot(this.raftId, (err, res) => {
+                    if (err) {
+                        this.log.error('error generating snapshot for ' +
+                        'ingestion', { err });
+                        return next(err);
+                    }
+                    batchState.logRes = { info: { start: 1 }, log: res };
+                    return next();
+                });
+            },
+        ], done);
+    }
+
+    _processLogEntry(batchState, record, entry, bucketdPrefix) {
+        // for a "del", entry.value will not exist but we still need to
+        // pass through the event
+        // for a bucket metadata entry from s3Connector, there will be no
+        // entry.key but there will be an entry.type and we
+        // need to pass this through
+        if (entry.key === undefined && entry.type === undefined) {
+            return;
+        }
+        if (!record.db) {
+            this._extensions.forEach(ext => ext.filter({
+                type: entry.type,
+                bucket: entry.bucket,
+                key: entry.key,
+                value: entry.value,
+            }));
+        } else {
+            if (record.db === 'users..bucket') {
+                const keySplit = entry.key.split('..|..');
+                entry.key =
+                    `${keySplit[0]}..|..${bucketdPrefix}-${keySplit[1]}`;
+            } else if (record.db === 'metastore') {
+                const keySplit = entry.key.split('/');
+                entry.key = `${keySplit[0]}/${bucketdPrefix}-${keySplit[1]}`;
+            } else {
+                if (record.db === entry.key) {
+                    entry.key = `${bucketdPrefix}-${entry.key}`;
+                }
+                record.db = `${bucketdPrefix}-${record.db}`;
+            }
+            this._extensions.forEach(ext => ext.filter({
+                type: entry.type,
+                bucket: record.db,
+                key: entry.key,
+                value: entry.value,
+            }));
+        }
+    }
+
+    _processPrepareEntries(batchState, done) {
+        const { entriesToPublish, logRes, logStats } = batchState;
+
+        this._setEntryBatch(entriesToPublish);
+
+        if (this.newIngestion) {
+            logRes.log.forEach(entry => {
+                logStats.nbLogRecordsRead += 1;
+                this._processLogEntry(batchState, entry, entry);
+            });
+            return done();
+        }
+        if (logRes.info.start === null || logRes.log === null) {
+            return done(null);
```

Comment:
There may be a hidden reason why we're passing `null` here instead of nothing, but eventually it should be rework to avoid having to pass `null` explicitly.